$MirOS: ports/www/dillo/stable/patches/patch-src_dillo_c,v 1.3 2006/06/24 18:05:03 bsiegert Exp $
$OpenBSD: patch-src_dillo_c,v 1.5 2004/02/21 01:37:19 couderc Exp $
--- src/dillo.c.orig	Mon Jan  2 19:19:45 2006
+++ src/dillo.c	Sun Jun 25 12:57:41 2006
@@ -49,6 +49,7 @@
 #include "interface.h"
 #include "dw.h"
 #include "cookies.h"
+#include "debug.h"
 
 
 /*
@@ -273,8 +274,8 @@ gint main(int argc, char *argv[])
    signal(SIGCHLD, Dillo_sigchld_handler);
 
 
-   /* check that ~/.dillo exists, create it if it doesn't */
-   dir = a_Misc_prepend_user_home(".dillo");
+   /* check that ~/.etc/dillo exists, create it if it doesn't */
+   dir = a_Misc_prepend_user_home(".etc/dillo");
    Dillo_check_home_dir(dir);
    g_free(dir);
 
@@ -349,15 +350,58 @@ gint main(int argc, char *argv[])
 }
 
 /*
- * Check if '~/.dillo' directory exists.
+ * This is from MirOS src/bin/mkdir/mkdir.c.
+ * mkpath -- create directories.
+ *	path     - path
+ *	mode     - file mode of terminal directory
+ *	dir_mode - file mode of intermediate directories
+ */
+int
+mkpath(char *path, mode_t mode, mode_t dir_mode)
+{
+	struct stat sb;
+	char *slash;
+	int done = 0;
+
+	slash = path;
+
+	while (!done) {
+		slash += strspn(slash, "/");
+		slash += strcspn(slash, "/");
+
+		done = (*slash == '\0');
+		*slash = '\0';
+
+		if (stat(path, &sb)) {
+			if (errno != ENOENT ||
+			    (mkdir(path, done ? mode : dir_mode) &&
+			    errno != EEXIST)) {
+				MSG("%s: %s", path, strerror(errno));
+				return (-1);
+			}
+		} else if (!S_ISDIR(sb.st_mode)) {
+			MSG("%s: %s", path, strerror(ENOTDIR));
+			return (-1);
+		}
+
+		*slash = '/';
+	}
+
+	return (0);
+}
+
+/*
+ * Check if '~/.etc/dillo' directory exists.
  * If not, try to create it.
  */
 static void Dillo_check_home_dir(char *dir)
 {
    struct stat st;
+   mode_t mode;
 
+   mode = (0777 & ~umask(0)) | S_IWUSR | S_IXUSR;
    if ( stat(dir, &st) == -1 ) {
-      if ( errno == ENOENT && mkdir(dir, 0700) < 0 ) {
+      if ( errno == ENOENT && mkpath(dir, 0700, mode) < 0 ) {
          MSG("Dillo: error creating directory %s: %s\n",
              dir, g_strerror(errno));
       } else
